File: problems/tier2-intermediate/016-amazon-locker/solution.cpp
#include <iostream>
#include <string>
#include <vector>
#include <map>
#include <queue>
#include <unordered_map>
using namespace std;

// ─── Data Model ─────────────────────────────────────────────────────────────

enum class LockerSize { SMALL, MEDIUM, LARGE };

struct Package {
    string packageId;
    LockerSize size;
};

struct Locker {
    string lockerId;
    LockerSize size;
    bool occupied;
};

// ─── Allocation Strategy ────────────────────────────────────────────────────

class LockerAllocationStrategy {
public:
    virtual string allocate(LockerSize packageSize,
                            map<LockerSize, queue<string>>& available) = 0;
    virtual ~LockerAllocationStrategy() = default;
};

class SmallestFitStrategy : public LockerAllocationStrategy {
public:
    string allocate(LockerSize packageSize,
                    map<LockerSize, queue<string>>& available) override {
        vector<LockerSize> tryOrder;
        if (packageSize == LockerSize::SMALL)
            tryOrder = {LockerSize::SMALL, LockerSize::MEDIUM, LockerSize::LARGE};
        else if (packageSize == LockerSize::MEDIUM)
            tryOrder = {LockerSize::MEDIUM, LockerSize::LARGE};
        else
            tryOrder = {LockerSize::LARGE};

        for (auto sz : tryOrder) {
            if (!available[sz].empty()) {
                string id = available[sz].front();
                available[sz].pop();
                return id;
            }
        }
        return "";
    }
};

// ─── Notification Channel ───────────────────────────────────────────────────

class NotificationChannel {
public:
    virtual void notify(const string& packageId, const string& message) = 0;
    virtual ~NotificationChannel() = default;
};

// ─── Deposit Record ─────────────────────────────────────────────────────────

struct DepositRecord {
    string lockerId;
    string packageId;
    string pickupCode;
    long   depositTime;
};

// ─── Locker System ──────────────────────────────────────────────────────────

class LockerSystem {
private:
    map<string, Locker> lockers;
    map<LockerSize, queue<string>> available;
    unordered_map<string, DepositRecord> activeDeposits;
    LockerAllocationStrategy* strategy;
    vector<NotificationChannel*> channels;
    int codeCounter = 0;
    int expiryHours = 0;

    string generateCode() {
        return "CODE-" + to_string(++codeCounter);
    }

    void notifyAll(const string& packageId, const string& message) {
        for (auto* ch : channels) {
            ch->notify(packageId, message);
        }
    }

    void freeLocker(const string& lockerId) {
        auto it = lockers.find(lockerId);
        if (it != lockers.end()) {
            it->second.occupied = false;
            available[it->second.size].push(lockerId);
        }
    }

public:
    LockerSystem() {
        strategy = new SmallestFitStrategy();
    }

    ~LockerSystem() {
        delete strategy;
    }

    void addLocker(const string& lockerId, LockerSize size) {
        lockers[lockerId] = {lockerId, size, false};
        available[size].push(lockerId);
    }

    string depositPackage(const string& packageId, LockerSize size,
                          long depositTime = 0) {
        string lockerId = strategy->allocate(size, available);
        if (lockerId.empty()) return "";

        lockers[lockerId].occupied = true;
        string code = generateCode();
        activeDeposits[code] = {lockerId, packageId, code, depositTime};
        notifyAll(packageId, "Package " + packageId + " deposited. Code: " + code);
        return code;
    }

    bool retrievePackage(const string& code) {
        auto it = activeDeposits.find(code);
        if (it == activeDeposits.end()) return false;

        freeLocker(it->second.lockerId);
        activeDeposits.erase(it);
        return true;
    }

    void setCodeExpiry(int hours) {
        expiryHours = hours;
    }

    vector<string> checkExpired(long currentTime) {
        vector<string> expired;
        if (expiryHours <= 0) return expired;

        for (auto it = activeDeposits.begin(); it != activeDeposits.end(); ) {
            if (currentTime - it->second.depositTime > (long)expiryHours * 3600) {
                freeLocker(it->second.lockerId);
                expired.push_back(it->second.packageId);
                notifyAll(it->second.packageId,
                          "Package " + it->second.packageId + " expired. Locker freed.");
                it = activeDeposits.erase(it);
            } else {
                ++it;
            }
        }
        return expired;
    }

    void addNotificationChannel(NotificationChannel* channel) {
        channels.push_back(channel);
    }
};

// ─── Global Entry Points ─────────────────────────────────────────────────────

static LockerSystem* g_system = nullptr;

void initLockerSystem() {
    delete g_system;
    g_system = new LockerSystem();
}

void addLocker(const string& lockerId, LockerSize size) {
    if (!g_system) initLockerSystem();
    g_system->addLocker(lockerId, size);
}

string depositPackage(const string& packageId, LockerSize size,
                      long depositTime = 0) {
    if (!g_system) initLockerSystem();
    return g_system->depositPackage(packageId, size, depositTime);
}

bool retrievePackage(const string& code) {
    if (!g_system) return false;
    return g_system->retrievePackage(code);
}

void setCodeExpiry(int hours) {
    if (!g_system) initLockerSystem();
    g_system->setCodeExpiry(hours);
}

vector<string> checkExpired(long currentTime) {
    if (!g_system) return {};
    return g_system->checkExpired(currentTime);
}

void addNotificationChannel(NotificationChannel* channel) {
    if (!g_system) initLockerSystem();
    g_system->addNotificationChannel(channel);
}

// ─── Ops simulator (used by spec-based tests) ──────────────────────────────

#include <memory>

static vector<string> g_chan_log;

class CapturingChannel : public NotificationChannel {
public:
    void notify(const string& packageId, const string& message) override {
        g_chan_log.push_back(packageId + ": " + message);
    }
};

struct LockerOp {
    string kind;
    string s1;
    string s2;
    int    i1;
    int    i2;
};

static LockerSize lsize_from(const string& s) {
    if (s == "S") return LockerSize::SMALL;
    if (s == "M") return LockerSize::MEDIUM;
    return LockerSize::LARGE;
}

vector<string> locker_simulate(vector<LockerOp> ops) {
    vector<string> out;
    vector<string> codes(32, "");
    static unique_ptr<CapturingChannel> chan;
    vector<string> last_expired;

    for (const auto& op : ops) {
        const string& k = op.kind;
        if (k == "new") {
            initLockerSystem();
            for (auto& c : codes) c.clear();
            g_chan_log.clear();
            chan.reset();
            last_expired.clear();
            out.push_back("ok");
        } else if (k == "add_locker") {
            addLocker(op.s1, lsize_from(op.s2));
            out.push_back("ok");
        } else if (k == "deposit") {
            string code = depositPackage(op.s1, lsize_from(op.s2), (long)op.i1);
            if (op.i2 >= 0 && op.i2 < (int)codes.size()) codes[op.i2] = code;
            out.push_back(code);
        } else if (k == "code_at") {
            out.push_back(codes[op.i2]);
        } else if (k == "retrieve") {
            out.push_back(retrievePackage(codes[op.i2]) ? "ok" : "fail");
        } else if (k == "retrieve_id") {
            out.push_back(retrievePackage(op.s1) ? "ok" : "fail");
        } else if (k == "set_expiry") {
            setCodeExpiry(op.i1);
            out.push_back("ok");
        } else if (k == "check_expired") {
            last_expired = checkExpired((long)op.i1);
            out.push_back(to_string((int)last_expired.size()));
        } else if (k == "expired_at") {
            out.push_back(op.i2 >= 0 && op.i2 < (int)last_expired.size() ? last_expired[op.i2] : "");
        } else if (k == "add_chan") {
            chan.reset(new CapturingChannel());
            addNotificationChannel(chan.get());
            out.push_back("ok");
        } else if (k == "chan_log_size") {
            out.push_back(to_string((int)g_chan_log.size()));
        } else if (k == "chan_log_contains") {
            bool found = false;
            for (auto& l : g_chan_log) if (l.find(op.s1) != string::npos) { found = true; break; }
            out.push_back(found ? "yes" : "no");
        } else {
            out.push_back("unknown:" + k);
        }
    }
    return out;
}

// ─── Main ────────────────────────────────────────────────────────────────────

#ifndef RUNNING_TESTS
int main() {
    initLockerSystem();

    addLocker("S1", LockerSize::SMALL);
    addLocker("S2", LockerSize::SMALL);
    addLocker("M1", LockerSize::MEDIUM);
    addLocker("L1", LockerSize::LARGE);

    string code1 = depositPackage("PKG001", LockerSize::SMALL);
    cout << "Deposited PKG001, code: " << code1 << endl;

    string code2 = depositPackage("PKG002", LockerSize::MEDIUM);
    cout << "Deposited PKG002, code: " << code2 << endl;

    bool ok = retrievePackage(code1);
    cout << "Retrieved PKG001: " << (ok ? "success" : "failed") << endl;

    return 0;
}
#endif

File: problems/tier2-intermediate/016-amazon-locker/solution.java
// Amazon Locker — Solution (Java)
import java.util.*;

class LockerOp {
    public String kind;
    public String s1;
    public String s2;
    public int i1;
    public int i2;

    public LockerOp(String kind, String s1, String s2, int i1, int i2) {
        this.kind = kind; this.s1 = s1; this.s2 = s2; this.i1 = i1; this.i2 = i2;
    }
}

enum LockerSize { SMALL, MEDIUM, LARGE }

class Locker {
    public String lockerId;
    public LockerSize size;
    public boolean occupied;
    public Locker(String id, LockerSize sz) { this.lockerId = id; this.size = sz; this.occupied = false; }
}

class DepositRecord {
    public String lockerId;
    public String packageId;
    public String pickupCode;
    public long depositTime;
    public DepositRecord(String l, String p, String c, long t) {
        lockerId = l; packageId = p; pickupCode = c; depositTime = t;
    }
}

interface LockerAllocationStrategy {
    String allocate(LockerSize packageSize, Map<LockerSize, Deque<String>> available);
}

class SmallestFitStrategy implements LockerAllocationStrategy {
    @Override
    public String allocate(LockerSize packageSize, Map<LockerSize, Deque<String>> available) {
        List<LockerSize> tryOrder = new ArrayList<>();
        if (packageSize == LockerSize.SMALL) {
            tryOrder.add(LockerSize.SMALL); tryOrder.add(LockerSize.MEDIUM); tryOrder.add(LockerSize.LARGE);
        } else if (packageSize == LockerSize.MEDIUM) {
            tryOrder.add(LockerSize.MEDIUM); tryOrder.add(LockerSize.LARGE);
        } else {
            tryOrder.add(LockerSize.LARGE);
        }
        for (LockerSize sz : tryOrder) {
            Deque<String> q = available.get(sz);
            if (q != null && !q.isEmpty()) return q.pollFirst();
        }
        return "";
    }
}

interface NotificationChannel {
    void notify(String packageId, String message);
}

class LockerSystem {
    Map<String, Locker> lockers = new LinkedHashMap<>();
    Map<LockerSize, Deque<String>> available = new EnumMap<>(LockerSize.class);
    Map<String, DepositRecord> activeDeposits = new LinkedHashMap<>();
    LockerAllocationStrategy strategy = new SmallestFitStrategy();
    List<NotificationChannel> channels = new ArrayList<>();
    int codeCounter = 0;
    int expiryHours = 0;

    public LockerSystem() {
        for (LockerSize s : LockerSize.values()) available.put(s, new ArrayDeque<>());
    }

    private String generateCode() {
        return "CODE-" + (++codeCounter);
    }

    private void notifyAll(String packageId, String message) {
        for (NotificationChannel ch : channels) ch.notify(packageId, message);
    }

    private void freeLocker(String lockerId) {
        Locker l = lockers.get(lockerId);
        if (l != null) {
            l.occupied = false;
            available.get(l.size).addLast(lockerId);
        }
    }

    public void addLocker(String lockerId, LockerSize size) {
        lockers.put(lockerId, new Locker(lockerId, size));
        available.get(size).addLast(lockerId);
    }

    public String depositPackage(String packageId, LockerSize size, long depositTime) {
        String lockerId = strategy.allocate(size, available);
        if (lockerId.isEmpty()) return "";
        lockers.get(lockerId).occupied = true;
        String code = generateCode();
        activeDeposits.put(code, new DepositRecord(lockerId, packageId, code, depositTime));
        notifyAll(packageId, "Package " + packageId + " deposited. Code: " + code);
        return code;
    }

    public boolean retrievePackage(String code) {
        DepositRecord rec = activeDeposits.get(code);
        if (rec == null) return false;
        freeLocker(rec.lockerId);
        activeDeposits.remove(code);
        return true;
    }

    public void setCodeExpiry(int hours) { this.expiryHours = hours; }

    public List<String> checkExpired(long currentTime) {
        List<String> expired = new ArrayList<>();
        if (expiryHours <= 0) return expired;
        Iterator<Map.Entry<String, DepositRecord>> it = activeDeposits.entrySet().iterator();
        while (it.hasNext()) {
            Map.Entry<String, DepositRecord> e = it.next();
            DepositRecord rec = e.getValue();
            if (currentTime - rec.depositTime > (long) expiryHours * 3600) {
                freeLocker(rec.lockerId);
                expired.add(rec.packageId);
                notifyAll(rec.packageId, "Package " + rec.packageId + " expired. Locker freed.");
                it.remove();
            }
        }
        return expired;
    }

    public void addNotificationChannel(NotificationChannel channel) {
        channels.add(channel);
    }
}

class CapturingChannel implements NotificationChannel {
    public List<String> log;
    public CapturingChannel(List<String> log) { this.log = log; }
    @Override
    public void notify(String packageId, String message) {
        log.add(packageId + ": " + message);
    }
}

public class Solution {
    private static LockerSystem system_instance = null;
    private static List<String> chanLog = new ArrayList<>();
    private static CapturingChannel currentChan = null;

    private static LockerSize lsizeFrom(String s) {
        if ("S".equals(s)) return LockerSize.SMALL;
        if ("M".equals(s)) return LockerSize.MEDIUM;
        return LockerSize.LARGE;
    }

    public static List<String> locker_simulate(List<LockerOp> ops) {
        List<String> out = new ArrayList<>();
        String[] codes = new String[32];
        for (int i = 0; i < 32; i++) codes[i] = "";
        List<String> lastExpired = new ArrayList<>();
        chanLog = new ArrayList<>();
        currentChan = null;
        system_instance = null;

        for (LockerOp op : ops) {
            String k = op.kind;
            if ("new".equals(k)) {
                system_instance = new LockerSystem();
                for (int i = 0; i < 32; i++) codes[i] = "";
                chanLog = new ArrayList<>();
                currentChan = null;
                lastExpired = new ArrayList<>();
                out.add("ok");
            } else if ("add_locker".equals(k)) {
                if (system_instance == null) system_instance = new LockerSystem();
                system_instance.addLocker(op.s1, lsizeFrom(op.s2));
                out.add("ok");
            } else if ("deposit".equals(k)) {
                if (system_instance == null) system_instance = new LockerSystem();
                String code = system_instance.depositPackage(op.s1, lsizeFrom(op.s2), (long) op.i1);
                if (op.i2 >= 0 && op.i2 < codes.length) codes[op.i2] = code;
                out.add(code);
            } else if ("code_at".equals(k)) {
                out.add(codes[op.i2]);
            } else if ("retrieve".equals(k)) {
                out.add(system_instance != null && system_instance.retrievePackage(codes[op.i2]) ? "ok" : "fail");
            } else if ("retrieve_id".equals(k)) {
                out.add(system_instance != null && system_instance.retrievePackage(op.s1) ? "ok" : "fail");
            } else if ("set_expiry".equals(k)) {
                if (system_instance == null) system_instance = new LockerSystem();
                system_instance.setCodeExpiry(op.i1);
                out.add("ok");
            } else if ("check_expired".equals(k)) {
                if (system_instance == null) {
                    lastExpired = new ArrayList<>();
                } else {
                    lastExpired = system_instance.checkExpired((long) op.i1);
                }
                out.add(Integer.toString(lastExpired.size()));
            } else if ("expired_at".equals(k)) {
                out.add(op.i2 >= 0 && op.i2 < lastExpired.size() ? lastExpired.get(op.i2) : "");
            } else if ("add_chan".equals(k)) {
                if (system_instance == null) system_instance = new LockerSystem();
                currentChan = new CapturingChannel(chanLog);
                system_instance.addNotificationChannel(currentChan);
                out.add("ok");
            } else if ("chan_log_size".equals(k)) {
                out.add(Integer.toString(chanLog.size()));
            } else if ("chan_log_contains".equals(k)) {
                boolean found = false;
                for (String l : chanLog) if (l.contains(op.s1)) { found = true; break; }
                out.add(found ? "yes" : "no");
            } else {
                out.add("unknown:" + k);
            }
        }
        return out;
    }
}
